src/mlpack/methods/det/dtree_impl.hpp

```diff
@@ -165,10 +165,150 @@ DTree<MatType, TagType>::DTree() :
+    else
+      left = NULL;
+    if(obj.right != NULL)
+      right = new DTree(*obj.right);
+    else
+      right = NULL;
+}
 
-// Root node initializers
+template <typename MatType, typename TagType>
+DTree<MatType, TagType>& DTree<MatType, TagType>::operator=(const DTree<MatType, TagType>& obj)
+{
+    //Copying the values from obj 
+    start = obj.start;
+    end = obj.end;
+    maxVals = obj.maxVals;
+    minVals = obj.minVals;
+    splitDim = obj.splitDim;
+    splitValue = obj.splitValue;
+    logNegError = obj.logNegError;
+    subtreeLeavesLogNegError = obj.subtreeLeavesLogNegError;
+    subtreeLeaves = obj.subtreeLeaves;
+    root = obj.root;
+    ratio = obj.ratio;
+    logVolume = obj.logVolume;
+    bucketTag = obj.bucketTag;
+    alphaUpper = obj.alphaUpper;
+    
+    //Copying the children   
+    if(obj.left != NULL)
+      left = new DTree(*obj.left);
+    else
+      left = NULL;
+    if(obj.right != NULL)
+      right = new DTree(*obj.right);
+    else
+      right = NULL;
+    
+    return *this;
+}
+
+template <typename MatType, typename TagType>
+DTree<MatType, TagType>::DTree(DTree&& obj):
+    start(obj.start),
+    end(obj.end),
+    maxVals(std::move(obj.maxVals)),
+    minVals(std::move(obj.minVals)),
+    splitDim(obj.splitDim),
+    splitValue(std::move(obj.splitValue)),
+    logNegError(obj.logNegError),
+    subtreeLeavesLogNegError(obj.subtreeLeavesLogNegError),
+    subtreeLeaves(obj.subtreeLeaves),
+    root(obj.root),
+    ratio(obj.ratio),
+    logVolume(obj.logVolume),
+    bucketTag(std::move(obj.bucketTag)),
+    alphaUpper(obj.alphaUpper),  
+    left(std::move(obj.left)),
+    right(std::move(obj.right))
+{   
+    //Set obj to default values
+    obj.start = 0;
+    obj.end = 0;
+    obj.splitDim = size_t(-1);
+    obj.splitValue = std::numeric_limits<ElemType>::max();
+    obj.logNegError = -DBL_MAX;
+    obj.subtreeLeavesLogNegError = -DBL_MAX;
+    obj.subtreeLeaves = 0;
+    obj.root = true;
+    obj.ratio = 1.0;
+    obj.logVolume = -DBL_MAX;
+    obj.bucketTag = -1;
+    obj.alphaUpper = 0.0;
+    obj.left = NULL;
+    obj.right = NULL;
+
+}
 
 template <typename MatType, typename TagType>
+DTree<MatType, TagType>& DTree<MatType, TagType>::operator=(DTree<MatType, TagType>&& obj)
+{
+    //Moving the values from object
+	start = obj.start;
+    end = obj.end;
+    splitDim = obj.splitDim;
+    logNegError = obj.logNegError;
+    subtreeLeavesLogNegError = obj.subtreeLeavesLogNegError;
+    subtreeLeaves = obj.subtreeLeaves;
+    root = obj.root;
+    ratio = obj.ratio;
+    logVolume = obj.logVolume;
+    alphaUpper = obj.alphaUpper;
+    maxVals = std::move(obj.maxVals);
+    minVals = std::move(obj.minVals);
+    splitValue = std::move(obj.splitValue);
+    bucketTag = std::move(obj.bucketTag);
+    
+    //Free the space allocated
+    delete left;
+    delete right;
+    
+    //Moving children
+    left = std::move(obj.left);
+    right = std::move(obj.right);
```

Comment:
Since you're just moving pointers here, you don't actually need to use `std::move()`, you can just do `left = obj.left` and `right = obj.right`.